project/MimaExcludes.scala

```diff
@@ -38,7 +38,11 @@ object MimaExcludes {
   lazy val v31excludes = v30excludes ++ Seq(
     // [SPARK-31077] Remove ChiSqSelector dependency on mllib.ChiSqSelectorModel
     // private constructor
-    ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.ml.feature.ChiSqSelectorModel.this")
+    ProblemFilters.exclude[IncompatibleMethTypeProblem]("org.apache.spark.ml.feature.ChiSqSelectorModel.this"),
+    // [SPARK-31127] Implement abstract Selector
+    ProblemFilters.exclude[MissingTypesProblem]("org.apache.spark.ml.feature.ChiSqSelectorModel"),
```

Comment:
Weird: this class is still there, right? is this still definitely a MiMa error? I'm trying to figure out if we're missing something here that causes an incompatible change.

This would only go into 3.1 right?
The refactoring looks OK; hard to evaluate it all.

Reply:
This is for 3.1 only.

The first two MiMa errors are because of type hierarchy change. 
```
[error]  * the type hierarchy of class org.apache.spark.ml.feature.ChiSqSelectorModel is different in current version. Missing types {org.apache.spark.ml.feature.ChiSqSelectorParams}
```
The 3rd MiMa error is due to return type change.
```
[error]  * method fit(org.apache.spark.sql.Dataset)org.apache.spark.ml.feature.ChiSqSelectorModel in class org.apache.spark.ml.feature.ChiSqSelector has a different result type in current version, where it is org.apache.spark.ml.feature.SelectorModel rather than org.apache.spark.ml.feature.ChiSqSelectorModel
```

Reply:
The first two seem OK; that won't cause an actual source or binary incompatibility (right?) just because there is an additional class in the hierarchy.

The third one doesn't sound quite right. The type should be ChiSqSelectorModel right? do you have to override it in the subclass like you did with methods in ChiSqSelectorModel? the superclass returns this.type which I thought was 'inherited' and would become ChiSqSelectorModel.

Am I right about that at all, and is there a workaround, or is MiMa incorrect somehow and the type is concretely ChiSqSelectorModel? or is this actually how lots of classes work?

Reply:
Seems I have to override ```fit``` in subclasses.